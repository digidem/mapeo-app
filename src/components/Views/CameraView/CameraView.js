// @flow
import React from 'react';
import { withNavigationFocus } from 'react-navigation';
import {
  Image,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
  Dimensions,
  ActivityIndicator
} from 'react-native';
import { RNCamera } from 'react-native-camera';
import I18n from 'react-native-i18n';
import type { Observation } from '../../../types/observation';
import { CHARCOAL, WHITE } from '../../../lib/styles.js';

import AddButton from '../../../images/add-button.png';

const styles = StyleSheet.create({
  cancelButton: {
    backgroundColor: CHARCOAL,
    justifyContent: 'center',
    height: 60,
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0
  },
  cancelText: {
    fontSize: 20,
    fontWeight: '700',
    textAlign: 'center',
    color: WHITE
  },
  capture: {
    alignSelf: 'center',
    marginTop: 375
  },
  preview: {
    flex: 1,
    alignItems: 'center'
  }
});

export type Props = {
  navigation: any,
  isFocused: boolean
};

export type StateProps = {
  selectedObservation: Observation
};

export type DispatchProps = {
  updateObservation: (o: Observation) => void,
  goToObservationEditor: () => void
};

type State = {
  loading: boolean
};

I18n.fallbacks = true;
I18n.translations = {
  en: require('../../../translations/en'),
  es: require('../../../translations/es')
};

class CameraView extends React.Component<
  Props & StateProps & DispatchProps,
  State
> {
  state = {
    loading: false
  };

  shouldComponentUpdate(
    nextProps: Props & StateProps & DispatchProps,
    nextState: State
  ) {
    if (nextProps.isFocused) {
      return nextProps !== this.props || nextState !== this.state;
    }

    return false;
  }

  camera: RNCamera;

  takePicture = async () => {
    const {
      updateObservation,
      selectedObservation,
      goToObservationEditor
    } = this.props;
    if (this.camera) {
      const options = { quality: 0.5, fixOrientation: true };
      try {
        this.setState({ loading: true });
        const data = await this.camera.takePictureAsync(options);
        this.setState({ loading: false });

        if (selectedObservation) {
          updateObservation({
            ...selectedObservation,
            media: selectedObservation.media.concat([
              {
                type: 'Photo',
                source: data.uri
              }
            ])
          });
          goToObservationEditor();
        }
      } catch (error) {
        console.warn(error);
      }
    }
  };

  render() {
    const { loading } = this.state;
    const { isFocused } = this.props;

    if (!isFocused) {
      return <View />;
    }

    return (
      <View style={{ flex: 1, flexDirection: 'column' }}>
        <RNCamera
          ref={ref => {
            this.camera = ref;
          }}
          style={styles.preview}
          type={RNCamera.Constants.Type.back}
          flashMode={RNCamera.Constants.FlashMode.off}
          permissionDialogTitle="Permission to use camera"
          permissionDialogMessage="We need your permission to use your camera phone"
        >
          <TouchableOpacity onPress={this.takePicture} style={styles.capture}>
            <Image
              source={AddButton}
              style={{
                width: 125,
                height: 125
              }}
            />
          </TouchableOpacity>
          <TouchableOpacity
            style={styles.cancelButton}
            onPress={() => this.props.navigation.goBack()}
          >
            <Text style={styles.cancelText}>Cancel</Text>
          </TouchableOpacity>
        </RNCamera>
        {loading && (
          <View
            style={{
              position: 'absolute',
              height: Dimensions.get('window').height - 30,
              width: Dimensions.get('window').width,
              backgroundColor: 'rgba(66,66,66,.8)',
              alignItems: 'center',
              justifyContent: 'center'
            }}
          >
            <View
              style={{
                backgroundColor: '#FFF',
                paddingHorizontal: 50,
                paddingVertical: 30,
                justifyContent: 'center',
                borderRadius: 3
              }}
            >
              <ActivityIndicator />
              <Text>{I18n.t('saving_image')}</Text>
            </View>
          </View>
        )}
      </View>
    );
  }
}

export default withNavigationFocus(CameraView);
