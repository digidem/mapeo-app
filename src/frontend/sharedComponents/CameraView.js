// @flow
import React from "react";
import { View, StyleSheet } from "react-native";
import { Camera } from "expo-camera";
import debug from "debug";
import { Accelerometer } from "expo-sensors";
import ImageResizer from "react-native-image-resizer";
import RNFS from "react-native-fs";

import AddButton from "./AddButton";
import { promiseTimeout } from "../lib/utils";
import bugsnag from "../lib/logger";
import type { CapturePromise } from "../context/DraftObservationContext";

const log = debug("CameraView");

const captureQuality = 75;
const captureOptions = {
  base64: false,
  exif: true,
  skipProcessing: true
};

type Props = {
  // Called when the user takes a picture, with a promise that resolves to an
  // object with the property `uri` for the captured (and rotated) photo.
  onAddPress: (e: any, capture: CapturePromise) => void
};

type State = {
  takingPicture: boolean,
  showCamera: boolean
};

type Acceleration = { x: number, y: number, z: number };

class CameraView extends React.Component<Props, State> {
  cameraRef: { current: any };
  subscription: { remove: () => any } | null;
  acceleration: Acceleration;
  mounted: boolean;
  state = { takingPicture: false, showCamera: true };

  constructor(props: Props) {
    super(props);
    this.cameraRef = React.createRef();
  }

  componentDidMount() {
    this.mounted = true;
    Accelerometer.isAvailableAsync().then(motionAvailable => {
      if (!motionAvailable || !this.mounted || this.subscription) return;
      Accelerometer.setUpdateInterval(1000);
      this.subscription = Accelerometer.addListener(acc => {
        this.acceleration = acc;
      });
    });
  }

  componentWillUnmount() {
    this.mounted = false;
    if (this.subscription) this.subscription.remove();
    this.subscription = null;
  }

  handleAddPress = (e: any) => {
    const camera = this.cameraRef.current;
    if (!camera) return log("Camera view not ready");
    if (this.state.takingPicture) return log("Shutter pressed twice");
    log("Start photo capture");
    const capture = promiseTimeout(
      camera.takePictureAsync(captureOptions),
      15000,
      "Error capturing photo"
    ).then(data => {
      log("Initial capture");
      return rotatePhoto(this.acceleration)(data);
    });
    this.setState(
      {
        takingPicture: true
      },
      () => {
        // Slight weirdness with a expo-camera bug: if we navigate away straight
        // away then the capture promise never resolves.
        setTimeout(this.props.onAddPress, 0, e, capture);
      }
    );
    capture.catch(bugsnag.notify);
    capture.finally(() => {
      if (!this.mounted) return;
      this.setState({ takingPicture: false });
    });
  };

  render() {
    const { takingPicture } = this.state;
    return (
      <View style={styles.container}>
        <Camera
          ref={this.cameraRef}
          style={{ flex: 1 }}
          type={Camera.Constants.Type.back}
          useCamera2Api={false}
        />
        <AddButton
          onPress={this.handleAddPress}
          style={{ opacity: takingPicture ? 0.5 : 1 }}
        />
      </View>
    );
  }
}

export default CameraView;

// Rotate the photo to match device orientation
function rotatePhoto(acc: Acceleration) {
  const rotation = getPhotoRotation(acc);
  return function({ uri, exif, width, height }) {
    const originalUri = uri;
    let resizedUri;
    const resizePromise = ImageResizer.createResizedImage(
      uri,
      width,
      height,
      "JPEG",
      captureQuality,
      rotation
    )
      .then(({ uri }) => {
        bugsnag.leaveBreadcrumb("Rotate photo", { type: "process" });
        // Image resizer uses `JPEG` as the extension, which gets passed through
        // to mapeo-core media store. Change to `jpg` to match legacy photos and
        // avoid issues on Windows (don't know if it recognizes `JPEG`)
        resizedUri = uri.replace(/\.JPEG$/, ".jpg");
        return RNFS.moveFile(uri, resizedUri);
      })
      .then(() => {
        bugsnag.leaveBreadcrumb("Rename photo", { type: "process" });
        RNFS.unlink(originalUri).then(() => log("Cleaned up un-rotated photo"));
        return { uri: resizedUri };
      })
      .catch(e => {
        bugsnag.notify(e, report => {
          report.errorMessage = "Error rotating photo";
          report.severity = "warning";
        });
        // Some devices throw an error trying to rotate the photo, so worst-case
        // we just don't rotate
        return { uri: originalUri, rotate: rotation };
      });
    return resizePromise;
  };
}

const ACC_AT_45_DEG = Math.sin(Math.PI / 4);

// Use the accelerometer to calculate the photo rotation, rotating as the user
// would expect based on the angle of the screen.
function getPhotoRotation(acc?: Acceleration) {
  if (!acc) return 0;
  const { x, y, z } = acc;
  let rotation = 0;
  if (z < -ACC_AT_45_DEG || z > ACC_AT_45_DEG) {
    // camera is pointing up or down
    if (Math.abs(y) > Math.abs(x)) {
      // camera is vertical
      if (y <= 0) rotation = 180;
      else rotation = 0;
    } else {
      // camera is horizontal
      if (x >= 0) rotation = -90;
      else rotation = 90;
    }
  } else if (x > -ACC_AT_45_DEG && x < ACC_AT_45_DEG) {
    // camera is vertical
    if (y <= 0) rotation = 180;
    else rotation = 0;
  } else {
    // camera is horizontal
    if (x >= 0) rotation = -90;
    else rotation = 90;
  }
  return rotation;
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "black"
  }
});
