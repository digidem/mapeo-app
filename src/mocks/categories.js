// @flow

export const createCategory = (category?: any) => ({
  id: 'categoryID',
  icon: require('../images/categories/category_18.png'),
  name: 'categoryName',
  fields: [],

  ...category
});
