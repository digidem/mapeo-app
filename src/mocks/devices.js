// @flow

export const createDevice = (device?: any) => ({
  id: '0',
  name: 'deviceName',
  selected: false,
  syncStatus: null,

  ...device
});
