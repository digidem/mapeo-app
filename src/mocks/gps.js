// @flow
import type { GPSState } from '../types/gps';
import {
  PERMISSION_DENIED,
  UNAVAILABLE,
  SEARCHING,
  LOW_ACCURACY,
  HIGH_ACCURACY
} from '../ducks/gps';

export const createGPSState = (partial: any): GPSState => ({
  status: HIGH_ACCURACY,
  coords: {
    longitude: 1,
    latitude: 1,
    accuracy: 50
  },
  timestamp: 0,

  ...partial
});
