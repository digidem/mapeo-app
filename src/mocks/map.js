// @flow
import type { Style } from '../types/map';

export const createStyle = (partial?: Object): Style => ({
  id: 'test-style',
  name: 'Test Style',
  bound: [1, 2, 3, 4],
  minzoom: 7,
  maxzoom: 8,

  ...partial
});
