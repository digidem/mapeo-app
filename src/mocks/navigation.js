import type { NavigationScreenProp } from 'react-navigation';

export const createNavigationScreenProp = () => {
  return {
    state: {},
    dispatch: jest.fn(),
    addListener: jest.fn(),
    getParam: jest.fn(),
    isFocused: jest.fn(),
    goBack: jest.fn(),
    navigate: jest.fn(),
    setParams: jest.fn()
  };
};
