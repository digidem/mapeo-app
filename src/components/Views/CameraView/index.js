// @flow
import { connect } from 'react-redux';
import type { Dispatch } from 'redux';
import {
  observationCreate,
  observationUpdate,
} from '../../../ducks/observations';
import { mediaSaveToCameraRoll } from '../../../ducks/media';
import { observationSource } from '../../../ducks/observationSource';
import type { StoreState, Action } from '../../../types/redux';
import CameraView from './CameraView';
import type { Props, StateProps, DispatchProps } from './CameraView';
import { drawerClose, drawerOpen } from '../../../ducks/drawers';

function mapStateToProps(state: StoreState, ownProps: Props): StateProps {
  return {
    observations: state.app.observations,
    selectedObservation: state.app.selectedObservation,
    showSavedModal: state.app.modals.saved,
    showEditorView:
      ownProps.navigation &&
      ownProps.navigation.state &&
      ownProps.navigation.state.params &&
      ownProps.navigation.state.params.showEditorView,
  };
}

function mapDispatchToProps(dispatch: Dispatch<any>): DispatchProps {
  return {
    createObservation: observation => dispatch(observationCreate(observation)),
    updateObservation: observation => dispatch(observationUpdate(observation)),
    onDrawerClose: () => dispatch(drawerClose('observations')),
    onDrawerOpen: () => dispatch(drawerOpen('observations')),
    updateObservationSource: () => dispatch(observationSource('camera')),
    saveToCameraRoll: meta => dispatch(mediaSaveToCameraRoll(meta)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(CameraView);
