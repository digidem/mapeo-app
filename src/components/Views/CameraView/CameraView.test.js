// @flow
import React from 'react';
import 'react-native';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import CameraView from './CameraView';
import { createNavigationScreenProp } from '../../../mocks/navigation';

jest.mock('../ObservationsView', () => () => null);
jest.mock('../MapView', () => () => null);
jest.mock('../../Base/SavedModal', () => () => null);

describe('CameraView tests', () => {
  const isFocused = () => true;
  const addListener = () => true;
  test('snapshots', () => {
    let tree;
    tree = shallow(
      <CameraView
        navigation={createNavigationScreenProp()}
        observations={{}}
        showSavedModal={false}
        showEditorView={false}
        createObservation={jest.fn()}
        updateObservation={jest.fn()}
        onDrawerClose={jest.fn()}
        onDrawerOpen={jest.fn()}
        updateObservationSource={jest.fn()}
        saveMedia={jest.fn()}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
