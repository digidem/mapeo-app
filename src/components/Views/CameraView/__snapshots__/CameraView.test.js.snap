// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CameraView tests snapshots 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CameraView
    createObservation={[MockFunction]}
    navigation={
      Object {
        "addListener": [MockFunction],
        "dispatch": [MockFunction],
        "getParam": [MockFunction],
        "goBack": [MockFunction],
        "isFocused": [MockFunction],
        "navigate": [MockFunction],
        "setParams": [MockFunction],
        "state": Object {},
      }
    }
    observations={Object {}}
    onDrawerClose={[MockFunction]}
    onDrawerOpen={[MockFunction]}
    saveMedia={[MockFunction]}
    showEditorView={false}
    showSavedModal={false}
    updateObservation={[MockFunction]}
    updateObservationSource={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "acceptDoubleTap": false,
      "acceptPan": true,
      "acceptPanOnDrawer": true,
      "acceptTap": false,
      "captureGestures": "open",
      "children": Array [
        <Unknown
          leftIcon={
            <TouchableOpacity
              activeOpacity={0.2}
              onPress={[Function]}
            >
              <Icon
                allowFontScaling={false}
                color="#ffffff"
                name="map"
                size={30}
              />
            </TouchableOpacity>
          }
          rightIcon={
            <TouchableOpacity
              activeOpacity={0.2}
              onPress={[Function]}
            >
              <Icon
                allowFontScaling={false}
                color="#ffffff"
                name="collections"
                size={30}
              />
            </TouchableOpacity>
          }
          style={
            Object {
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": 5,
            }
          }
        />,
        <Unknown />,
        <Component
          style={
            Object {
              "flex": 1,
              "flexDirection": "column",
            }
          }
        >
          <RNCamera
            permissionDialogMessage="We need your permission to use your camera phone"
            permissionDialogTitle="Permission to use camera"
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
              }
            }
            type="back"
          >
            <TouchableOpacity
              activeOpacity={0.2}
              onPress={[Function]}
              style={
                Object {
                  "alignSelf": "center",
                  "bottom": 2,
                  "position": "absolute",
                }
              }
            >
              <Image
                source={
                  Object {
                    "testUri": "../../../src/images/add-button.png",
                  }
                }
                style={
                  Object {
                    "height": 125,
                    "width": 125,
                  }
                }
              />
            </TouchableOpacity>
          </RNCamera>
        </Component>,
      ],
      "closedDrawerOffset": 0,
      "content": <Unknown
        closeRightDrawer={[Function]}
        navigation={
          Object {
            "addListener": [MockFunction],
            "dispatch": [MockFunction],
            "getParam": [MockFunction],
            "goBack": [MockFunction],
            "isFocused": [MockFunction],
            "navigate": [MockFunction],
            "setParams": [MockFunction],
            "state": Object {},
          }
        }
      />,
      "disabled": false,
      "elevation": 0,
      "initializeOpen": false,
      "negotiatePan": false,
      "onClose": [Function],
      "onCloseStart": [MockFunction],
      "onOpen": [Function],
      "onOpenStart": [MockFunction],
      "open": null,
      "openDrawerOffset": 0,
      "panCloseMask": null,
      "panOpenMask": null,
      "panThreshold": 0.25,
      "side": "right",
      "styles": Object {},
      "tapToClose": false,
      "tweenDuration": 250,
      "tweenEasing": "linear",
      "tweenHandler": null,
      "type": "displace",
      "useInteractionManager": false,
    },
    "ref": [Function],
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "leftIcon": <TouchableOpacity
            activeOpacity={0.2}
            onPress={[Function]}
          >
            <Icon
              allowFontScaling={false}
              color="#ffffff"
              name="map"
              size={30}
            />
          </TouchableOpacity>,
          "rightIcon": <TouchableOpacity
            activeOpacity={0.2}
            onPress={[Function]}
          >
            <Icon
              allowFontScaling={false}
              color="#ffffff"
              name="collections"
              size={30}
            />
          </TouchableOpacity>,
          "style": Object {
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 5,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <RNCamera
              permissionDialogMessage="We need your permission to use your camera phone"
              permissionDialogTitle="Permission to use camera"
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                }
              }
              type="back"
            >
              <TouchableOpacity
                activeOpacity={0.2}
                onPress={[Function]}
                style={
                  Object {
                    "alignSelf": "center",
                    "bottom": 2,
                    "position": "absolute",
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../src/images/add-button.png",
                    }
                  }
                  style={
                    Object {
                      "height": 125,
                      "width": 125,
                    }
                  }
                />
              </TouchableOpacity>
            </RNCamera>,
            false,
          ],
          "style": Object {
            "flex": 1,
            "flexDirection": "column",
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <TouchableOpacity
                  activeOpacity={0.2}
                  onPress={[Function]}
                  style={
                    Object {
                      "alignSelf": "center",
                      "bottom": 2,
                      "position": "absolute",
                    }
                  }
                >
                  <Image
                    source={
                      Object {
                        "testUri": "../../../src/images/add-button.png",
                      }
                    }
                    style={
                      Object {
                        "height": 125,
                        "width": 125,
                      }
                    }
                  />
                </TouchableOpacity>,
                false,
              ],
              "flashMode": undefined,
              "permissionDialogMessage": "We need your permission to use your camera phone",
              "permissionDialogTitle": "Permission to use camera",
              "style": Object {
                "alignItems": "center",
                "flex": 1,
              },
              "type": "back",
            },
            "ref": [Function],
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "activeOpacity": 0.2,
                  "children": <Image
                    source={
                      Object {
                        "testUri": "../../../src/images/add-button.png",
                      }
                    }
                    style={
                      Object {
                        "height": 125,
                        "width": 125,
                      }
                    }
                  />,
                  "onPress": [Function],
                  "style": Object {
                    "alignSelf": "center",
                    "bottom": 2,
                    "position": "absolute",
                  },
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "source": Object {
                      "testUri": "../../../src/images/add-button.png",
                    },
                    "style": Object {
                      "height": 125,
                      "width": 125,
                    },
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
              false,
            ],
            "type": [Function],
          },
          false,
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "acceptDoubleTap": false,
        "acceptPan": true,
        "acceptPanOnDrawer": true,
        "acceptTap": false,
        "captureGestures": "open",
        "children": Array [
          <Unknown
            leftIcon={
              <TouchableOpacity
                activeOpacity={0.2}
                onPress={[Function]}
              >
                <Icon
                  allowFontScaling={false}
                  color="#ffffff"
                  name="map"
                  size={30}
                />
              </TouchableOpacity>
            }
            rightIcon={
              <TouchableOpacity
                activeOpacity={0.2}
                onPress={[Function]}
              >
                <Icon
                  allowFontScaling={false}
                  color="#ffffff"
                  name="collections"
                  size={30}
                />
              </TouchableOpacity>
            }
            style={
              Object {
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 5,
              }
            }
          />,
          <Unknown />,
          <Component
            style={
              Object {
                "flex": 1,
                "flexDirection": "column",
              }
            }
          >
            <RNCamera
              permissionDialogMessage="We need your permission to use your camera phone"
              permissionDialogTitle="Permission to use camera"
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                }
              }
              type="back"
            >
              <TouchableOpacity
                activeOpacity={0.2}
                onPress={[Function]}
                style={
                  Object {
                    "alignSelf": "center",
                    "bottom": 2,
                    "position": "absolute",
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../src/images/add-button.png",
                    }
                  }
                  style={
                    Object {
                      "height": 125,
                      "width": 125,
                    }
                  }
                />
              </TouchableOpacity>
            </RNCamera>
          </Component>,
        ],
        "closedDrawerOffset": 0,
        "content": <Unknown
          closeRightDrawer={[Function]}
          navigation={
            Object {
              "addListener": [MockFunction],
              "dispatch": [MockFunction],
              "getParam": [MockFunction],
              "goBack": [MockFunction],
              "isFocused": [MockFunction],
              "navigate": [MockFunction],
              "setParams": [MockFunction],
              "state": Object {},
            }
          }
        />,
        "disabled": false,
        "elevation": 0,
        "initializeOpen": false,
        "negotiatePan": false,
        "onClose": [Function],
        "onCloseStart": [MockFunction],
        "onOpen": [Function],
        "onOpenStart": [MockFunction],
        "open": null,
        "openDrawerOffset": 0,
        "panCloseMask": null,
        "panOpenMask": null,
        "panThreshold": 0.25,
        "side": "right",
        "styles": Object {},
        "tapToClose": false,
        "tweenDuration": 250,
        "tweenEasing": "linear",
        "tweenHandler": null,
        "type": "displace",
        "useInteractionManager": false,
      },
      "ref": [Function],
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "leftIcon": <TouchableOpacity
              activeOpacity={0.2}
              onPress={[Function]}
            >
              <Icon
                allowFontScaling={false}
                color="#ffffff"
                name="map"
                size={30}
              />
            </TouchableOpacity>,
            "rightIcon": <TouchableOpacity
              activeOpacity={0.2}
              onPress={[Function]}
            >
              <Icon
                allowFontScaling={false}
                color="#ffffff"
                name="collections"
                size={30}
              />
            </TouchableOpacity>,
            "style": Object {
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": 5,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <RNCamera
                permissionDialogMessage="We need your permission to use your camera phone"
                permissionDialogTitle="Permission to use camera"
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                  }
                }
                type="back"
              >
                <TouchableOpacity
                  activeOpacity={0.2}
                  onPress={[Function]}
                  style={
                    Object {
                      "alignSelf": "center",
                      "bottom": 2,
                      "position": "absolute",
                    }
                  }
                >
                  <Image
                    source={
                      Object {
                        "testUri": "../../../src/images/add-button.png",
                      }
                    }
                    style={
                      Object {
                        "height": 125,
                        "width": 125,
                      }
                    }
                  />
                </TouchableOpacity>
              </RNCamera>,
              false,
            ],
            "style": Object {
              "flex": 1,
              "flexDirection": "column",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <TouchableOpacity
                    activeOpacity={0.2}
                    onPress={[Function]}
                    style={
                      Object {
                        "alignSelf": "center",
                        "bottom": 2,
                        "position": "absolute",
                      }
                    }
                  >
                    <Image
                      source={
                        Object {
                          "testUri": "../../../src/images/add-button.png",
                        }
                      }
                      style={
                        Object {
                          "height": 125,
                          "width": 125,
                        }
                      }
                    />
                  </TouchableOpacity>,
                  false,
                ],
                "flashMode": undefined,
                "permissionDialogMessage": "We need your permission to use your camera phone",
                "permissionDialogTitle": "Permission to use camera",
                "style": Object {
                  "alignItems": "center",
                  "flex": 1,
                },
                "type": "back",
              },
              "ref": [Function],
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "activeOpacity": 0.2,
                    "children": <Image
                      source={
                        Object {
                          "testUri": "../../../src/images/add-button.png",
                        }
                      }
                      style={
                        Object {
                          "height": 125,
                          "width": 125,
                        }
                      }
                    />,
                    "onPress": [Function],
                    "style": Object {
                      "alignSelf": "center",
                      "bottom": 2,
                      "position": "absolute",
                    },
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "source": Object {
                        "testUri": "../../../src/images/add-button.png",
                      },
                      "style": Object {
                        "height": 125,
                        "width": 125,
                      },
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  "type": [Function],
                },
                false,
              ],
              "type": [Function],
            },
            false,
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
